spec/services/course/user_invitation_service_spec.rb

```diff
@@ -0,0 +1,228 @@
+      end
+    end
+    let(:new_user_attributes) do
+      new_users.map do |user|
+        { name: user.name, email: user.email }
+      end
+    end
+    let(:users) { existing_users + new_users }
+    let(:user_attributes) { existing_user_attributes + new_user_attributes }
+
+    describe '#invite' do
+      def verify_users
+        new_users.each do |user|
+          course_user = course.course_users.find do |course_user|
+            course_user.invitation && course_user.invitation.user_email.email == user.email
+          end
+          expect(course_user).not_to be_nil
+          expect(course_user).to be_invited
+          expect(course_user.name).to eq(user.name)
+        end
+        existing_users.each do |user|
+          course_user = course.course_users.find do |course_user|
+            course_user.user && course_user.user.email == user.email
+          end
+          expect(course_user).not_to be_nil
+          expect(course_user).to be_approved
+          expect(course_user.name).to eq(user.name)
+        end
+      end
+
+      it 'accepts a list of invitation attributes' do
+        subject.invite(user_attributes)
+      end
+
+      it 'accepts a CSV file with a header' do
+        subject.invite(temp_csv_from_attributes(user_attributes.map do |attributes|
+          OpenStruct.new(attributes)
+        end))
+
+        verify_users
+      end
+    end
+
+    describe '#invite_from_file' do
+      subject do
+        Course::UserInvitationService.new(course).tap do |result|
+          result.define_singleton_method(:invite_users) do |users|
+            users
+          end
+        end
+      end
+      let(:temp_csv) { temp_csv_from_attributes(users) }
+      after { temp_csv.close! }
+
+      it 'accepts a file with name/header' do
+        result = subject.send(:invite_from_file, temp_csv)
+        expect(result.length).to eq(users.length)
+      end
+
+      it 'calls #invite_users with appropriate user attributes' do
+        result = subject.send(:invite_from_file, temp_csv)
+        expect(result).to eq(user_attributes)
+      end
+    end
+
+    describe '#invite_users' do
+      context 'when users already exist in the current instance' do
+        it 'immediately adds users' do
+          subject.send(:invite_users, existing_user_attributes)
+          existing_users.each do |user|
+            found_user = course.course_users.find { |course_user| course_user.user == user }
+            expect(found_user).not_to be_nil
+          end
+        end
+      end
+
+      context 'when users do not exist in the current instance' do
+        it 'invites users' do
+          subject.send(:invite_users, new_user_attributes)
+          new_users.each do |user|
+            expect(course.course_users.any? do |course_user|
+                course_user.invitation.present? &&
+                  course_user.invitation.user_email.email == user.email &&
+                  course_user.invited?
+              end).to be_truthy
```

Comment:
`end` at 118, 14 is not aligned with `course.course_users.any? do |course_user|` at 114, 19 or `expect(course.course_users.any? do |course_user|` at 114, 12